src/main/java/io/vertx/core/http/impl/Http2UpgradedClientConnection.java

```diff
@@ -228,25 +233,51 @@ public void channelRead(ChannelHandlerContext ctx, Object msg) throws Exception
             // Therefore we must buffer all pending messages until the request is fully sent.
 
             int maxContent = maxContentLength();
-            boolean lower = pendingSize < maxContent;
+            boolean lower = bufferedSize < maxContent;
             if (msg instanceof ByteBufHolder) {
-              pendingSize += ((ByteBufHolder)msg).content().readableBytes();
+              bufferedSize += ((ByteBufHolder)msg).content().readableBytes();
             } else if (msg instanceof ByteBuf) {
-              pendingSize += ((ByteBuf)msg).readableBytes();
+              bufferedSize += ((ByteBuf)msg).readableBytes();
             }
+            buffered.add(msg);
 
-            if (pendingSize >= maxContent) {
-              if (lower) {
-                pending.clear();
-                ctx.fireExceptionCaught(new TooLongFrameException("Max content exceeded " + maxContentLength() + " bytes."));
-              }
-              return;
+            if (bufferedSize >= maxContent && lower) {
+              ctx.fireExceptionCaught(new TooLongFrameException("Max content exceeded " + maxContentLength() + " bytes."));
             }
-            pending.add(msg);
           } else {
             super.channelRead(ctx, msg);
           }
         }
+
+        @Override
+        public void userEventTriggered(ChannelHandlerContext ctx, Object evt) throws Exception {
+          if (SEND_BUFFERED_MESSAGES == evt) {
+            Deque<Object> messages = buffered;
+            buffered = null;
+            Object msg;
+            while ((msg = messages.poll()) != null) {
+              super.channelRead(ctx, msg);
+            }
+          } else {
+            super.userEventTriggered(ctx, evt);
+          }
+        }
+
+        @Override
+        public void handlerRemoved(ChannelHandlerContext ctx) throws Exception {
+          if (buffered != null) {
+            Deque<Object> messages = buffered;
+            buffered = null;
```

Comment:
why not reusing `buffered` insteadd of nulling it out?

Reply:
because it is used as a marker by `channelRead`

Reply:
and in practice if that's not null then it means that something bad happen